File: android/core/src/main/kotlin/com/sangwook/ptimer/core/catalog/CustomFilmReferenceTableResolver.kt
package com.sangwook.ptimer.core.catalog

import com.sangwook.ptimer.core.reciprocity.TableAnchor

/**
 * Resolves the linked reference-table anchors for a saved custom FORMULA
 * film from its persisted `UserEditableMetadata.referenceTableFilmID`
 * (PTIMER-180). Display-only: the anchors drive reference/error rendering
 * and are NEVER fed into calculation. Mirrors iOS
 * `CustomFilmReferenceTableResolver`.
 */
object CustomFilmReferenceTableResolver {

    data class Resolution(
        /** Current anchors of the linked table, or empty when unlinked or unresolvable. */
        val anchors: List<TableAnchor>,
        /** True when a link exists but no longer resolves to a custom table. */
        val isLinkedButMissing: Boolean,
    )

    /**
     * Resolves the reference table for [formulaFilm]. [lookup] maps a film id
     * to the current [FilmIdentity] in the library. An unlinked formula
     * returns an empty, not-missing resolution.
     */
    fun resolve(formulaFilm: FilmIdentity, lookup: (String) -> FilmIdentity?): Resolution {
        val tableId = formulaFilm.userMetadata?.referenceTableFilmID
            ?: return Resolution(emptyList(), isLinkedButMissing = false)
        val table = lookup(tableId) ?: return Resolution(emptyList(), isLinkedButMissing = true)
        val anchors = tableAnchors(table)
        return if (anchors.isEmpty()) {
            Resolution(emptyList(), isLinkedButMissing = true)
        } else {
            Resolution(anchors, isLinkedButMissing = false)
        }
    }

    private fun tableAnchors(film: FilmIdentity): List<TableAnchor> {
        val rules = film.profiles.firstOrNull()?.typedRules ?: return emptyList()
        return rules.firstNotNullOfOrNull { (it as? ReciprocityRule.Table)?.rule?.anchors } ?: emptyList()
    }
}

File: ios/PTimerKit/Sources/PTimerKit/CustomFilm/CustomFilmReferenceTableResolver.swift
// Copyright © 2026 Sangwook Han
// SPDX-License-Identifier: Apache-2.0

import Foundation
import PTimerCore

/// PTIMER-180: resolves the linked reference-table anchors for a saved
/// custom **formula** film from its persisted
/// `UserEditableMetadata.referenceTableFilmID`.
///
/// The creation flow already holds the table anchors in memory; the
/// **edit** flow opens from persisted metadata and must re-hydrate them
/// so the formula editor's graph markers and the Calculation Basis
/// Reference / Error columns reappear.
///
/// Display-only: the anchors drive comparison / error rendering and are
/// never fed into calculation. Errors recompute from the table's
/// *current* anchors, so a table edited after the formula was saved is
/// reflected on the next open without touching the formula parameters.
public enum CustomFilmReferenceTableResolver {

    public struct Resolution: Equatable {
        /// Current anchors of the linked table, or empty when the
        /// formula is unlinked or the link cannot be resolved.
        public let anchors: [TableAnchor]
        /// `true` when the formula carries a `referenceTableFilmID` but
        /// it no longer resolves to a custom table (deleted, or not a
        /// table) — the editor then shows "reference table unavailable"
        /// rather than reference / error columns.
        public let isLinkedButMissing: Bool

        public init(anchors: [TableAnchor], isLinkedButMissing: Bool) {
            self.anchors = anchors
            self.isLinkedButMissing = isLinkedButMissing
        }
    }

    /// Resolves the reference table for `formulaFilm`. `lookup` maps a
    /// film id to the current `FilmIdentity` in the library (so an
    /// edited table surfaces its latest anchors). Unlinked formulas
    /// return an empty, not-missing resolution so the editor behaves
    /// exactly like the pre-PTIMER-180 unlinked editor.
    public static func resolve(
        for formulaFilm: FilmIdentity,
        lookup: (String) -> FilmIdentity?
    ) -> Resolution {
        guard let tableID = formulaFilm.userMetadata?.referenceTableFilmID else {
            return Resolution(anchors: [], isLinkedButMissing: false)
        }
        guard let table = lookup(tableID) else {
            return Resolution(anchors: [], isLinkedButMissing: true)
        }
        let anchors = tableAnchors(of: table)
        return anchors.isEmpty
            ? Resolution(anchors: [], isLinkedButMissing: true)
            : Resolution(anchors: anchors, isLinkedButMissing: false)
    }

    private static func tableAnchors(of film: FilmIdentity) -> [TableAnchor] {
        for rule in film.profiles.first?.rules ?? [] {
            if case let .tableInterpolation(table) = rule {
                return table.anchors
            }
        }
        return []
    }
}
